modules/auxiliary/gather/ldap_esc_vulnerable_cert_finder.rb

```diff
@@ -328,106 +337,289 @@ def find_esc3_vuln_cert_templates
+    # Obtain the authenticated user information to check if they have write permissions on the certificate templates
     begin
-      whoami_response = @ldap.ldapwhoami
-    rescue Net::LDAP::Error => e
-      print_warning("The module failed to run the ldapwhoami command, ESC4 detection can't continue. Error was: #{e.class}: #{e.message}.")
+      authenticated_user_info = get_authenticated_user_info
+    rescue LdapWhoamiError => e
+      print_warning("ESC4 detection skipped: #{e.message}")
       return
     end
 
-    if whoami_response.empty?
-      print_error("Unable to retrieve the username using ldapwhoami, ESC4 detection can't continue")
-      return
+    esc_raw_filter = '(objectclass=pkicertificatetemplate)'
+    attributes = ['cn', 'description', 'ntSecurityDescriptor']
+    esc_entries = query_ldap_server(esc_raw_filter, attributes, base_prefix: CERTIFICATE_TEMPLATES_BASE)
+
+    return if esc_entries.empty?
+
+    esc_entries.each do |entry|
+      certificate_symbol = entry[:cn][0].to_sym
+      next if @certificate_details[certificate_symbol][:enroll_sids].empty?
+
+      security_descriptor = Rex::Proto::MsDtyp::MsDtypSecurityDescriptor.read(entry[:ntsecuritydescriptor].first)
+      if user_can_write?(authenticated_user_info, security_descriptor)
+        @certificate_details[certificate_symbol][:techniques] << 'ESC4'
+        @certificate_details[certificate_symbol][:notes] << "ESC4: The account: #{authenticated_user_info[:samaccountname].first} has edit permissions over the template #{certificate_symbol}."
+      end
     end
+  end
 
-    sam_account_name = whoami_response.split('\\')[1]
-    user_raw_filter = "(sAMAccountName=#{sam_account_name})"
-    attributes = ['DN', 'objectSID', 'objectClass', 'primarygroupID']
-    our_account = query_ldap_server(user_raw_filter, attributes)&.first
-    if our_account.nil?
-      print_warning("Unable to determine the User SID for #{sam_account_name}, ESC4 detection can't continue")
-      return
+  def get_object_by_dn(dn)
+    object = @ldap_objects.find { |o| o['dn']&.first == dn }
+    return object if object
+
+    object = query_ldap_server("(distinguishedName=#{ldap_escape_filter(dn)})", nil)&.first
+    @ldap_objects << object if object
+    object
+  end
+
+  def get_object_by_samaccountname(samaccountname)
+    object = @ldap_objects.find { |o| o['sAMAccountName']&.first == samaccountname }
+
+    if object.nil?
+      object = query_ldap_server("(sAMAccountName=#{ldap_escape_filter(samaccountname)})", nil)&.first
+      @ldap_objects << object if object
     end
 
-    user_sid = map_sids_to_names([Rex::Proto::MsDtyp::MsDtypSid.read(our_account[:objectsid].first).value]).first
-    domain_sid = user_sid.value.to_s.rpartition('-').first
-    user_groups = []
+    object
+  end
+
+  def get_authenticated_user_info
+    # Check if the authenticated user info is already cached
+    cached_user_info = @ldap_objects.find { |obj| obj[:type] == :authenticated_user }
```

Comment:
Related to my other comment about storing things that aren't LDAP objects in the `@ldap_objects` array seems counter-intuitive. This also seems like we probably don't even need to cache it. If `@whoami` is caching the name of the current user, then the LDAP object can just be cached and stored with the other LDAP objects by it's `sAMAccountName` attribute.